wpimath/src/main/java/edu/wpi/first/math/trajectory/Trajectory.java

```diff
@@ -1,150 +1,93 @@
+public abstract class Trajectory<SampleType extends TrajectorySample<SampleType>> {
 
-  private final double m_totalTime;
-  private final List<State> m_states;
+  /** The samples this Trajectory is composed of. */
+  public final SampleType[] samples;
 
-  /** Constructs an empty trajectory. */
-  public Trajectory() {
-    m_states = new ArrayList<>();
-    m_totalTime = 0.0;
-  }
+  private final InterpolatingTreeMap<Time, SampleType> sampleMap;
+
+  /** The total duration of the trajectory. */
+  @JsonIgnore public final Time duration;
 
   /**
-   * Constructs a trajectory from a vector of states.
+   * Constructs a Trajectory.
    *
-   * @param states A vector of states.
-   * @throws IllegalArgumentException if the vector of states is empty.
+   * @param samples the samples of the trajectory. Order does not matter as they will be ordered
+   *     internally.
    */
-  public Trajectory(final List<State> states) {
-    m_states = states;
-
-    if (m_states.isEmpty()) {
-      throw new IllegalArgumentException("Trajectory manually created with no states.");
+  @SuppressWarnings({"this-escape"})
+  public Trajectory(SampleType[] samples) {
+    this.samples =
+        Arrays.stream(samples)
+            .sorted(Comparator.comparingDouble(s -> s.timestamp.in(Seconds)))
+            .toArray(size -> Arrays.copyOf(samples, size));
+
+    this.sampleMap =
+        new InterpolatingTreeMap<>(
+            (start, end, q) ->
+                MathUtil.inverseLerp(start.in(Seconds), end.in(Seconds), q.in(Seconds)),
+            this::interpolate);
+
+    for (var sample : this.samples) {
+      sampleMap.put(sample.timestamp, sample);
     }
 
-    m_totalTime = m_states.get(m_states.size() - 1).time;
+    this.duration =
+        this.samples.length > 0 ? this.samples[this.samples.length - 1].timestamp : Seconds.of(0.0);
   }
 
   /**
-   * Linearly interpolates between two poses.
+   * Interpolates between two samples. This method must be implemented by subclasses to provide
+   * drivetrain-specific interpolation logic.
    *
-   * @param startValue The start pose.
-   * @param endValue The end pose.
-   * @param t The fraction for interpolation.
-   * @return The interpolated pose.
+   * @param start The starting sample.
+   * @param end The ending sample.
+   * @param t The interpolation parameter between 0 and 1.
+   * @return The interpolated sample.
    */
-  private static Pose2d lerp(Pose2d startValue, Pose2d endValue, double t) {
-    return startValue.plus((endValue.minus(startValue)).times(t));
-  }
+  public abstract SampleType interpolate(SampleType start, SampleType end, double t);
 
-  /**
-   * Returns the initial pose of the trajectory.
-   *
-   * @return The initial pose of the trajectory.
-   */
-  public Pose2d getInitialPose() {
-    return sample(0).pose;
+  /** Gets the first sample in the trajectory. */
+  public SampleType start() {
+    return sampleAt(0.0);
```

Comment:
This is more efficient than a map lookup.
```suggestion
    return samples[0];
```